pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/BucketDelayedDeliveryTracker.java

```diff
@@ -0,0 +1,569 @@
+                    deliverAt - clock.millis());
+        }
+
+        updateTimer();
+
+        checkAndUpdateHighest(deliverAt);
+
+        return true;
+    }
+
+    @Override
+    public synchronized boolean hasMessageAvailable() {
+        long cutoffTime = getCutoffTime();
+
+        boolean hasMessageAvailable = getNumberOfDelayedMessages() > 0 && nextDeliveryTime() <= cutoffTime;
+        if (!hasMessageAvailable) {
+            updateTimer();
+        }
+        return hasMessageAvailable;
+    }
+
+    @Override
+    protected long nextDeliveryTime() {
+        if (getPriorityQueue().isEmpty() && !sharedBucketPriorityQueue.isEmpty()) {
+            return sharedBucketPriorityQueue.peekN1();
+        } else if (sharedBucketPriorityQueue.isEmpty() && !getPriorityQueue().isEmpty()) {
+            return getPriorityQueue().peekN1();
+        }
+        long timestamp = getPriorityQueue().peekN1();
+        long bucketTimestamp = sharedBucketPriorityQueue.peekN1();
+        return Math.min(timestamp, bucketTimestamp);
+    }
+
+    @Override
+    public synchronized long getNumberOfDelayedMessages() {
+        return numberDelayedMessages;
+    }
+
+    @Override
+    public synchronized long getBufferMemoryUsage() {
+        return getPriorityQueue().bytesCapacity() + sharedBucketPriorityQueue.bytesCapacity();
+    }
+
+    @Override
+    public synchronized Set<PositionImpl> getScheduledMessages(int maxMessages) {
+        long cutoffTime = getCutoffTime();
+
+        moveScheduledMessageToSharedQueue(cutoffTime);
+
+        Set<PositionImpl> positions = new TreeSet<>();
+        int n = maxMessages;
+
+        while (n > 0 && !sharedBucketPriorityQueue.isEmpty()) {
+            long timestamp = sharedBucketPriorityQueue.peekN1();
+            if (timestamp > cutoffTime) {
+                break;
+            }
+
+            long ledgerId = sharedBucketPriorityQueue.peekN2();
+            long entryId = sharedBucketPriorityQueue.peekN3();
+            positions.add(new PositionImpl(ledgerId, entryId));
+
+            sharedBucketPriorityQueue.pop();
+            removeIndexBit(ledgerId, entryId);
+
+            BucketState bucketState = snapshotSegmentLastIndexTable.remove(ledgerId, entryId);
+            if (bucketState != null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] Load next snapshot segment, bucketState: {}", dispatcher.getName(), bucketState);
+                }
+                // All message of current snapshot segment are scheduled, load next snapshot segment
+                // TODO make it asynchronous and not blocking this process
+                try {
+                    asyncLoadNextBucketSnapshotEntry(bucketState, false).get(AsyncOperationTimeoutSeconds,
+                            TimeUnit.SECONDS);
+                } catch (InterruptedException | ExecutionException | TimeoutException e) {
+                    throw new RuntimeException(e);
```

Comment:
If an exception occurs here, the next segment will never load again.

Reply:
I will improve it in the future by make it asynchronous.

Reply:
Should we add a `TODO` here?

Reply:
Ok.